enhancements/authentication/synchronizing-user-groups-on-login.md

```diff
@@ -0,0 +1,235 @@
+are stored in a 3rd party identity provider system. When OpenShift retrieves
+such an identity, it usually only gathers information about the user's unique
+identification in that IdP and optionally a username that's preferred for the
+user if it's different from its identifier.
+
+It is a common practice to group users in logical groups on the identity provider
+side and sometimes it would make sense to bring the groups over to the OpenShift
+cluster as well. The identity providers oftentimes allow a protocol-specific path
+to retrieve group information, which is something OpenShift should leverage in those
+cases where it is desired.
+
+### Goals
+
+- allow retrieving user's groups from a 3rd party OIDC identity provider into
+  OpenShift when such a user logs in
+
+### Non-Goals
+
+- implement group synchronization for an identity provider type other than
+  OIDC
+- keep the groups retrieved from a 3rd party identity provider synchronized
+  between user logins
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+I am an administrator of an OpenShift cluster that uses identities from
+my company's identity provider. The users of the identity provider are
+organized in a complex and well-though-out group structure. I would like
+OpenShift to respect the group structure so that I can assign privileges
+to the users on OpenShift side according to their group membership.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### OIDC Group Retrieval
+
+OIDC core specification does not mention a standard `id_token` claim that
+would serve for user group storage and so none of the standard scopes define
+accessing it. That means retrieving the groups will very likely always require
+the user to configure `extraScopes` in their identity provider configuration
+(oauth.config.openshift.io/cluster).
+
+A possible follow-up implementation might include being able to
+[request specific claims](https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims)
+from the UserInfo endpoint or in an `id_token`, which would allow for a more fine-grained
+configuration to reach the group claim.
+
+#### API
+
+The `OpenIDClaims` config API structure shall be expanded by a new `Groups`
+field:
+
+```golang
+// OpenIDClaims contains a list of OpenID claims to use when authenticating with an OpenID identity provider
+type OpenIDClaims struct {
+    ...
+
+    // groups is the list of claims whose values should be converted into OpenShift groups
+    // that will include the user
+    // +optional
+    Groups []string `json:"groups,omitempty"`
+}
+```
+
+The group object should be annotated with `oauth.openshift.io/idpname: <idpname>`
+where `<idpname>` marks the identity provider name from the configuration. It should
+also get an `oauth.openshift.io/generated: true` if it hasn't previously existed.
+
+#### oauth-server
+
+When the OIDC provider has group synchronization configured, the oauth-server
```

Comment:
"configured" means implictely:
1. an `extraScopes` entry is configured (optional)
2. a non-empty set of `Groups` entries is configured
?

Reply:
2. should be enough